### PR TITLE
refactor!: Using the MsgEthereumTx interface in the rpc module

### LINKED ISSUE #853: [Refactor] Using the MsgEthereumTx interface in the rpc module
I understand that cosmos/evm will continue to follow up on compatibility with ethereum.
As the structure of MsgEthereumTx was modified to introduce EIP7702, the above case may occur again in the future.
I think that if we define MsgEthereumTx, which is used in the rpc module like the ante module's GaxTx, as an interface, we can consider future expandability.

Although we're not to preserve ethermint's backwards compatibility, this approach could solve that problem as well.
The above ethermint related content is related to the hardcoded `*evmtypes.MsgEthereumTx` that was discussed in Slack.

### PR DESCRIPTION
**Pull request by nowooj**
# Description

This code doesn't cover all MsgEthereumTx implementations within the rpc module, but it does cover the query part. This code can be expanded and structured into a completely independent code structure.

[Removing Hash()](https://github.com/cosmos/evm/commit/6d023bb4002886584197a3ddaa35590a21134eec) hasn't important to do with the PR, but it was added for backward compatibility since ethermint uses Hash as a variable.

Closes: #853

---

## Author Checklist

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

I have...

- [ ] tackled an existing issue or discussed with a team member
- [ ] left instructions on how to review the changes
- [x] targeted the `main` branch

### CODE REVIEW

File: ante/tx_listener.go [hunk 1]
```diff
@@ -27,7 +27,7 @@ func (d TxListenerDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate boo
 	if ctx.IsCheckTx() && !simulate && d.pendingTxListener != nil {
 		for _, msg := range tx.GetMsgs() {
 			if ethTx, ok := msg.(*evmtypes.MsgEthereumTx); ok {
```
Comment (vladjdk) [inline ante/tx_listener.go:29]: Why are we leaving some *evmtypes.MsgEthereumTx as pointers instead of the interface?

Comment (nowooj) [reply ante/tx_listener.go:29]: I thought that newly execute tx, for example tx coming through ante handler or mempool, would always use the latest evmtypes.MsgEthereumTx

**Review by vladjdk — CHANGES_REQUESTED** [review]
Lint and conflicts

**Comment by vladjdk** [comment]
@greptile review

File: rpc/backend/comet_to_eth.go [hunk 2]
```diff
@@ -292,8 +294,8 @@ func (b *Backend) ReceiptsFromCometBlock(
 		}
 
 		contractAddress := common.Address{}
-		if ethMsg.Raw.To() == nil {
-			contractAddress = crypto.CreateAddress(ethMsg.GetSender(), ethMsg.Raw.Nonce())
+		if tx.To() == nil {
+			contractAddress = crypto.CreateAddress(common.Address(ethMsg.GetFrom().Bytes()), tx.Nonce())
```
[the remark this answers is not in the document]
Comment (nowooj) [reply rpc/backend/comet_to_eth.go]: https://github.com/cosmos/evm/pull/854/commits/cd50ea2f5bceaf07d79a6fff36180f851ddb78e2

File: x/vm/types/msg.go [hunk 3]
```diff
@@ -13,6 +13,7 @@ import (
 	protov2 "google.golang.org/protobuf/proto"
 
 	evmapi "github.com/cosmos/evm/api/cosmos/evm/vm/v1"
+	"github.com/cosmos/evm/rpc/types/interfaces"
```
[the remark this answers is not in the document]
Comment (nowooj) [reply x/vm/types/msg.go]: https://github.com/cosmos/evm/pull/854/commits/56544ac8bdf1099dc477f5bfe2150f7f99ef7efb

File: rpc/types/interfaces/interface.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	FromEthereumTx(tx *ethtypes.Transaction)
+	FromSignedEthereumTx(tx *ethtypes.Transaction, signer ethtypes.Signer) error
+	GetFrom() sdk.AccAddress
+	GetGas() uint64
+	GetEffectiveFee(baseFee *big.Int) *big.Int
+	AsTransaction() *ethtypes.Transaction
+	GetSenderLegacy(signer ethtypes.Signer) (common.Address, error)
+	AsMessage(baseFee *big.Int) *core.Message
+	UnmarshalBinary(b []byte, signer ethtypes.Signer) error
+}
```
[the remark this answers is not in the document]
Comment (nowooj) [reply rpc/types/interfaces/interface.go]: https://github.com/cosmos/evm/pull/854/commits/40ace1edeab01bbf304c68f829efa684d09f53d4

**Review by aljo242 — CHANGES_REQUESTED** [review]
**ci has not run:** fork pr, all workflows `action_required`. also 34 commits behind main.

**`Hash()` removal is a bundled breaking change:** the pr description acknowledges this was kept for ethermint backward compat. removing it breaks any downstream calling `.Hash()` directly on `*MsgEthereumTx`. move this to a separate pr with an explicit changelog entry.

**undocumented behavioral change in contract address derivation:** `comet_to_eth.go:299` changes `ethMsg.GetSender()` (recovered from signature) to `ethMsg.GetFrom()` (stored field). these should be equivalent for well-formed txs, but the difference isn't documented and could silently produce wrong contract addresses for edge cases. needs a comment or test confirming equivalence.

**`GetSenderLegacy` in interface but unused:** no call site in this diff uses it through the interface -- either document why it's required or remove it.

**Comment by nowooj** [comment]
> **ci has not run:** fork pr, all workflows `action_required`. also 34 commits behind main.
> 
> **`Hash()` removal is a bundled breaking change:** the pr description acknowledges this was kept for ethermint backward compat. removing it breaks any downstream calling `.Hash()` directly on `*MsgEthereumTx`. move this to a separate pr with an explicit changelog entry.
> 
> **undocumented behavioral change in contract address derivation:** `comet_to_eth.go:299` changes `ethMsg.GetSender()` (recovered from signature) to `ethMsg.GetFrom()` (stored field). these should be equivalent for well-formed txs, but the difference isn't documented and could silently produce wrong contract addresses for edge cases. needs a comment or test confirming equivalence.
> 
> **`GetSenderLegacy` in interface but unused:** no call site in this diff uses it through the interface -- either document why it's required or remove it.

@aljo242 Thanks, addressed the actionable parts.

- Added an `API-BREAKING` changelog entry for the `MsgEthereumTx.Hash()` removal and documented `AsTransaction().Hash()` as the replacement.
- Updated contract creation receipt address derivation to use `GetSenderLegacy(signer)` instead of `GetFrom()` directly, preserving the fallback sender recovery for older decoded txs with an empty `From`.
- `GetSenderLegacy` is used through `RPCMsgEthereumTxI` in RPC/query paths (`getAccountNonce`, `GetTransactionReceipt`, `GetBlockReceipts`), so I kept it in the interface.
- The branch has also been updated with latest `main`.
